### PR TITLE
MONIT-18683: Fixed issue with start time of traces generated to file.

### PR DESCRIPTION
**Pull request by dbaghdasarya**

### CODE REVIEW

File: src/main/java/com/wavefront/datastructures/TraceFromWF.java [hunk 1]
```diff
@@ -1,26 +1,79 @@
...
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;
 import com.fasterxml.jackson.databind.ObjectMapper;
 
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.UUID;
 
+@JsonAutoDetect(fieldVisibility = Visibility.ANY)
```
Comment (sirak92) [inline src/main/java/com/wavefront/datastructures/TraceFromWF.java:21]: I would like to see javadoc here describing how it's different from Trace object. Also as a general comment, all public methods should have javadoc.

File: src/main/java/com/wavefront/datastructures/TraceFromWF.java [hunk 2]
```diff
@@ -1,26 +1,79 @@
...
 
-  public String traceId;
-  public long start_ms;
-  public long end_ms;
-  public long total_duration_ms;
-  public List<SpanFromWF> spans;
+  private String traceId;
+  private long start_ms = Long.MAX_VALUE;
+  private long end_ms = Long.MIN_VALUE;
+  private long total_duration_ms;
```
Comment (sirak92) [inline src/main/java/com/wavefront/datastructures/TraceFromWF.java]: By java naming convention the fields should be in camel case. Is there any reason to name them differently. I see it in several places as well. Also, the traceId and total_duration_ms wasn't used at all.

File: src/main/java/com/wavefront/datastructures/Trace.java [hunk 3]
```diff
@@ -128,13 +128,9 @@ public long getTotal_duration_ms() {
```
Comment (sirak92) [inline src/main/java/com/wavefront/datastructures/Trace.java]: Span, WFSpan and <Trace, TraceWF> pairs have a lot of common parts, so maybe makes sense to separate out a abstract class from them?

Comment (sirak92) [reply src/main/java/com/wavefront/datastructures/Trace.java]: One more thing related to WFSpan. I see that it's not part of this review, but I would like to change public fields to private with getters/setters

File: src/main/java/com/wavefront/datastructures/TraceFromWF.java [hunk 4]
```diff
@@ -1,26 +1,79 @@
...
+  private List<SpanFromWF> spans;
+
+  public List<SpanFromWF> getSpans() {
+    return spans;
+  }
+
+  public void setSpans(List<SpanFromWF> spans) {
+    this.spans = spans;
+    setStartAndEnd();
+    if (this.spans == null || this.spans.size() == 0) {
```
Comment (sirak92) [inline src/main/java/com/wavefront/datastructures/TraceFromWF.java]: please use spans.isEmpty() instaed

File: src/main/java/com/wavefront/datastructures/TraceFromWF.java [hunk 5]
```diff
@@ -1,26 +1,79 @@
...
+    });
+    total_duration_ms = end_ms - start_ms;
+  }
 
   public void shiftTrace(long deltaMillis) {
+    start_ms = Long.MAX_VALUE;
+    end_ms = Long.MIN_VALUE;
     if (spans == null) {
       return;
     }
```
Comment (sirak92) [inline src/main/java/com/wavefront/datastructures/TraceFromWF.java]: move it to the beginning of the function, please

File: src/main/java/com/wavefront/datastructures/TraceFromWF.java [hunk 6]
```diff
@@ -1,26 +1,79 @@
...
+  private long total_duration_ms;
+  private List<SpanFromWF> spans;
+
+  public List<SpanFromWF> getSpans() {
+    return spans;
+  }
+
+  public void setSpans(List<SpanFromWF> spans) {
+    this.spans = spans;
+    setStartAndEnd();
```
Comment (sirak92) [inline src/main/java/com/wavefront/datastructures/TraceFromWF.java]: I guess the is needed only for non-null(non-empty) case, so could you move it under else statement.

File: src/main/java/com/wavefront/datastructures/TraceFromWF.java [hunk 7]
```diff
@@ -1,26 +1,79 @@
...
+  }
+
+  public long getTotal_duration_ms() {
+    return end_ms - start_ms;
+  }
+
+  private void setStartAndEnd() {
+    if (spans == null) {
+      return;
+    }
```
Comment (sirak92) [inline src/main/java/com/wavefront/datastructures/TraceFromWF.java]: you can add empty checks for all similar cases, to not do any additional stuff.

File: src/main/java/com/wavefront/datastructures/TraceFromWF.java [hunk 8]
```diff
@@ -1,26 +1,79 @@
...
+      traceId = null;
+    } else {
+      traceId = spans.get(0).traceId;
+    }
+  }
+
+  public long getStart_ms() {
+    if (start_ms == Long.MAX_VALUE) {
+      setStartAndEnd();
+    }
```
Comment (sirak92) [inline src/main/java/com/wavefront/datastructures/TraceFromWF.java]: the setStartAndEnd() function is called after setSpans(), so I think these call is redundant

File: src/main/java/com/wavefront/datastructures/TraceFromWF.java [hunk 9]
```diff
@@ -1,26 +1,79 @@
...
+    return start_ms;
+  }
+
+  public long getEnd_ms() {
+    if (end_ms == Long.MIN_VALUE) {
+      setStartAndEnd();
+    }
```
Comment (sirak92) [inline src/main/java/com/wavefront/datastructures/TraceFromWF.java]: same here

File: src/main/java/com/wavefront/generators/FromPatternGenerator.java [hunk 10]
```diff
@@ -73,11 +76,25 @@ public FromPatternGenerator(GeneratorConfig config, DataQueue dataQueue) {
...
 
-      dataQueue.addTrace(generateTrace(traceTypePattern));
+      while (generatedSpans < mustBeGeneratedSpans) {
+        // get next trace type to be generated
+        TraceTypePattern traceTypePattern = getNextTraceType(generatorConfig.getTraceTypePatterns());
+        Trace trace = generateTrace(traceTypePattern,
+            (currentSeconds - random.nextInt(SLEEP_DELAY_SECONDS)) * 1000);
+        dataQueue.addTrace(trace);
+        generatedSpans += trace.getSpansCount();
+      }
```
Comment (sirak92) [inline src/main/java/com/wavefront/generators/FromPatternGenerator.java]: now this function become very similar to run() implementation, so could you make one parameterized function to not have code duplication.

File: src/main/java/com/wavefront/datastructures/TraceFromWF.java (hunk 8 above)
Comment (dbaghdasarya) [reply src/main/java/com/wavefront/datastructures/TraceFromWF.java]: Agree

**Pull request merged**